### ISSUE TITLE
XGB model prediction feature validation check

### ISSUE THREAD

**Issue by EvanzzzZ** [dmlc/xgboost#3074]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System: Mac

Compiler: 

Package used (python/R/jvm/C++):

`xgboost` version used: 0.6

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`)
2. Logs will be helpful (If logs are large, please upload as attachment).

If you are using jvm package, please 

1. add [jvm-packages] in the title to make it quickly be identified
2. the gcc version and distribution

If you are using python package, please provide

1. The python version and distribution
2. The command to install `xgboost` if you are not installing from source

If you are using R package, please provide

1. The R `sessionInfo()`
2. The command to install `xgboost` if you are not installing from source

I found some strange things about XGB model prediction.
I am not sure if it's a but, so I would like to put it here for discussion.

Here is a sample code:
(Assuming I already my model ready)
```
dtest1 = xgb.DMatrix("test1")
model = xgb.Booster()
model.load_model('xgboost-model')
preds1 = model.predict(dtest1)
print preds1

dtest2 = xgb.DMatrix("test2")
preds2 = model.predict(dtest2)
print preds2
```

I loaded the model and do prediction with two different dataset. 
Assuming test1 is:
`1 1:0.1 20:0.4`
test2 is:
`1 1:0.1 10:0.4`

XGB will complain feature mismatch when doing scoring for test2.

Switch the order of test1 and test2, XGB will complain the same thing for scoring of test1.

Basically XGB model initiated its num of features from the first scoring data it received,
and then keep that as a fixed standard to verify all the upcoming scoring input.

Is this the case ?
Please correct me if I was wrong somewhere.

**Comment by hcho3** [comment]
Fixed in #3323.
